ext/drreg/drreg.c

```diff
@@ -1719,37 +2714,82 @@ drreg_event_restore_state(void *drcontext, bool restore_memory,
+            instr_free(drcontext, &inst);
+            instr_free(drcontext, &next_inst);
+            return true;
+        }
+        decode(drcontext, pc, &next_inst);
+        if (is_our_spill_or_restore(drcontext, &inst, &next_inst, &is_spill, &reg, &slot,
+                                    &offs, &is_indirect_spill)) {
             LOG(drcontext, DR_LOG_ALL, 3,
                 "%s @" PFX " found %s to %s offs=0x%x => slot %d\n", __FUNCTION__,
-                prev_pc, spill ? "spill" : "restore", get_register_name(reg), offs, slot);
-            if (spill) {
+                prev_pc, is_spill ? "is_spill" : "restore", get_register_name(reg), offs,
+                slot);
+            if (is_spill) {
                 if (slot == AFLAGS_SLOT) {
                     spilled_to_aflags = slot;
+#ifdef SIMD_SUPPORTED
+                } else if (is_indirect_spill) {
+                    if (reg_is_vector_simd(reg)) {
+                        ASSERT(slot < ops.num_spill_simd_slots, "slots is out-of-bounds");
+                        if (spilled_simd_to[SIMD_IDX(reg)] < MAX_SIMD_SPILLS &&
+                            /* allow redundant spill */
+                            spilled_simd_to[SIMD_IDX(reg)] != slot) {
+                            /* This reg is already spilled: we assume that this new
+                             * spill is to a tmp slot for preserving the tool's value.
+                             */
+                            LOG(drcontext, DR_LOG_ALL, 3,
+                                "%s @" PFX ": ignoring tool is_spill\n", __FUNCTION__,
+                                pc);
+                        } else {
+                            spilled_simd_to[SIMD_IDX(reg)] = slot;
+                            simd_slot_use[slot] = reg;
+                        }
+                    }
+#endif
                 } else if (spilled_to[GPR_IDX(reg)] < MAX_SPILLS &&
-                           /* allow redundant spill */
+                           /* allow redundant is_spill */
                            spilled_to[GPR_IDX(reg)] != slot) {
                     /* This reg is already spilled: we assume that this new spill
                      * is to a tmp slot for preserving the tool's value.
                      */
-                    LOG(drcontext, DR_LOG_ALL, 3, "%s @" PFX ": ignoring tool spill\n",
+                    LOG(drcontext, DR_LOG_ALL, 3, "%s @" PFX ": ignoring tool is_spill\n",
                         __FUNCTION__, pc);
                 } else {
                     spilled_to[GPR_IDX(reg)] = slot;
                 }
             } else {
-                if (slot == AFLAGS_SLOT && spilled_to_aflags == slot)
+                /* Not a spill, but a restore. */
+                if (slot == AFLAGS_SLOT && spilled_to_aflags == slot) {
                     spilled_to_aflags = MAX_SPILLS;
-                else if (spilled_to[GPR_IDX(reg)] == slot)
+#ifdef SIMD_SUPPORTED
+                } else if (is_indirect_spill) {
```

Comment:
Since this is used both for indirect spill and restore, rename accordingly, e.g. just is_indirect?

Reply:
Done